test/e2e/network_segmentation.go

```diff
@@ -365,8 +372,226 @@ var _ = Describe("Network Segmentation", func() {
 			),
 		)
 	})
+
+	Context("Manage user-defined-network with UserDefinedNetwork CRD", func() {
+		const (
+			testUdnName                = "test-net"
+			userDefinedNetworkResource = "userdefinednetwork"
+		)
+
+		BeforeEach(func() {
+			By("create tests UserDefinedNetwork")
+			cleanup, err := createManifest(f.Namespace.Name, newUserDefinedNetworkManifest(testUdnName))
+			DeferCleanup(cleanup)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(waitForUserDefinedNetworkReady(f.Namespace.Name, testUdnName, 5*time.Second)).To(Succeed())
+		})
+
+		It("should create NetworkAttachmentDefinition according to spec", func() {
+			udnUidRaw, err := e2ekubectl.RunKubectl(f.Namespace.Name, "get", userDefinedNetworkResource, testUdnName, "-o", "jsonpath='{.metadata.uid}'")
+			Expect(err).NotTo(HaveOccurred(), "should get the UserDefinedNetwork UID")
+			testUdnUID := strings.Trim(udnUidRaw, "'")
+
+			By("verify a NetworkAttachmentDefinition is created according to spec")
+			assertNetAttachDefManifest(nadClient, f.Namespace.Name, testUdnName, testUdnUID)
+		})
+
+		It("should delete NetworkAttachmentDefinition when UserDefinedNetwork is deleted", func() {
+			By("delete UserDefinedNetwork")
+			_, err := e2ekubectl.RunKubectl(f.Namespace.Name, "delete", userDefinedNetworkResource, testUdnName)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("verify a NetworkAttachmentDefinition has been deleted")
+			Eventually(func() bool {
+				_, err := nadClient.NetworkAttachmentDefinitions(f.Namespace.Name).Get(context.Background(), testUdnName, metav1.GetOptions{})
+				return err != nil && kerrors.IsNotFound(err)
+			}, time.Second*3, time.Second*1).Should(BeTrue(),
+				"NetworkAttachmentDefinition should be deleted following UserDefinedNetwork deletion")
+		})
+
+		Context("pod connected to UserDefinedNetwork", func() {
+			const testPodName = "test-pod-udn"
+
+			var (
+				udnInUseDeleteTimeout = 65 * time.Second
+				deleteNetworkTimeout  = 5 * time.Second
+				deleteNetworkInterval = 1 * time.Second
+			)
+
+			BeforeEach(func() {
+				By("create pod")
+				networkAttachments := []nadapi.NetworkSelectionElement{
+					{Name: testUdnName, Namespace: f.Namespace.Name},
+				}
+				cfg := podConfig(testPodName, withNetworkAttachment(networkAttachments))
+				cfg.namespace = f.Namespace.Name
+				runUDNPod(cs, f.Namespace.Name, *cfg, nil)
+			})
+
+			It("UserDefinedNetwork cannot be deleted when being used", func() {
+				By("verify UserDefinedNetwork cannot be deleted")
+				cmd := e2ekubectl.NewKubectlCommand(f.Namespace.Name, "delete", userDefinedNetworkResource, testUdnName)
+				cmd.WithTimeout(time.NewTimer(deleteNetworkTimeout).C)
+				_, err := cmd.Exec()
+				Expect(err).To(HaveOccurred(),
+					"should fail to delete UserDefinedNetwork when used")
+
+				By("verify UserDefinedNetwork associated NetworkAttachmentDefinition cannot be deleted")
```

Comment:
right, what I want to check here is not the format of the error message, but the the reason for UDN not being deleted. I don't think we need to build the full error message, as it may change over time, but just using something like `strings.Contains()` may be good enough to find pod name in the error message.
Now I am thinking it may useful to have this as a separate reason, different from "SyncError", but we can work on it in the follow up. Status check can then also be postponed, as you like